### PR TITLE
PKI Module Update

### PR DESCRIPTION
**Pull request by wa5i**
1.Optimized code within the PKI module.
2.Enable PKI module to generate root CA certificates.

### CODE REVIEW

File: src/modules/pki/field.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            description: r#"The requested common name; if you want more than
+one, specify the alternative names in the alt_names map. If not specified when
+signing, the common name will be taken from the CSR; other names must still be
+specified in alt_names or ip_sans.
+"#
+        },
+        "ttl": {
+            field_type: FieldType::Str,
+            description: r#"The requested Time To Live for the certificate;
+sets the expiration date. If not specified the role default, backend default,
```
Comment (InfoHunter) [inline src/modules/pki/field.rs:31]: What is the difference between TTL and expiration date?

Comment (wa5i) [reply src/modules/pki/field.rs:31]: > What is the difference between TTL and expiration date?

Prioritize using the 'not_after' parameter; if not configured, use 'ttl'.

**Comment by wa5i** [comment]
ping @InfoHunter

**Comment by InfoHunter** [comment]
Very busy on other things, wait a couple of days...

File: src/modules/pki/mod.rs [hunk 2]
```diff
@@ -807,6 +315,8 @@ x/+V28hUf8m8P2NxP5ALaDZagdaMfzjGZo3O3wDv33Cds0P5GMGQYnRXDxcZN/2L
             "locality": "HZ",
             "organization": "ANT-Group",
             "ou": "Big-Security",
+            "street_address": "Xixi Road",
+            "postal_code": "30010",
```
Comment (InfoHunter) [inline src/modules/pki/mod.rs:319]: Change this organization information to OpenAtom, something like: CN/Beijing/Beijing/OpenAtom/....

File: src/modules/pki/mod.rs [hunk 3]
```diff
@@ -822,15 +332,19 @@ x/+V28hUf8m8P2NxP5ALaDZagdaMfzjGZo3O3wDv33Cds0P5GMGQYnRXDxcZN/2L
...
+        assert_eq!(role_data["not_before_duration"].as_u64().unwrap(), 30);
         assert_eq!(role_data["key_type"].as_str().unwrap(), "rsa");
         assert_eq!(role_data["key_bits"].as_u64().unwrap(), 4096);
         assert_eq!(role_data["country"].as_str().unwrap(), "CN");
         assert_eq!(role_data["province"].as_str().unwrap(), "ZJ");
         assert_eq!(role_data["locality"].as_str().unwrap(), "HZ");
         assert_eq!(role_data["organization"].as_str().unwrap(), "ANT-Group");
         assert_eq!(role_data["ou"].as_str().unwrap(), "Big-Security");
+        assert_eq!(role_data["street_address"].as_str().unwrap(), "Xixi Road");
+        assert_eq!(role_data["postal_code"].as_str().unwrap(), "30010");
```
Comment (InfoHunter) [inline src/modules/pki/mod.rs:347]: Same as above

File: src/modules/pki/path_keys.rs [hunk 4]
```diff
@@ -1,16 +1,218 @@
...
+                    required: true,
+                    field_type: FieldType::Int,
+                    description: r#"
+The number of bits to use. Allowed values are 0 (universal default); with rsa
+key_type: 2048 (default), 3072, or 4096; with ec key_type: 224, 256 (default),
+384, or 521; ignored with ed25519."#
+                },
+                "key_type": {
+                    field_type: FieldType::Str,
+                    default: "rsa",
```
Comment (InfoHunter) [inline src/modules/pki/path_keys.rs:42]: Is this lowercase only?

File: src/modules/pki/mod.rs (hunk 2 above)
Comment (wa5i) [reply src/modules/pki/mod.rs]: > Change this organization information to OpenAtom, something like: CN/Beijing/Beijing/OpenAtom/....

fixed

File: src/modules/pki/mod.rs (hunk 3 above)
Comment (wa5i) [reply src/modules/pki/mod.rs]: > Same as above

fixed

File: src/modules/pki/path_keys.rs (hunk 4 above)
Comment (wa5i) [reply src/modules/pki/path_keys.rs:42]: > Is this lowercase only?

Case-insensitive, fixed

**Review by InfoHunter — APPROVED** [review]

**Pull request merged**